### PR TITLE
[ISSUE-635] Support WBT disabling

### LINKED ISSUE #635: Support custom WBT settings pass
**Is your feature request related to a problem? Please describe.**
CSI should have an opportunity to set custom WBT value for device.

**Describe the solution you'd like**
Options:
- disable/enable for kernel versions range
- restore default value

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by safronovD**
### Resolves #635

Changes described in proposal:
- Add watcher for Wbt configmap to update it every 60 sec
- On NodeStage: try to read and edit `wbt_lat_usec`
- On NodeUnstage: restore default value for `wbt_lat_usec`

## PR checklist
- [x] Add link to the issue
- [x] Choose Project
- [x] Choose PR label
- [x] New unit tests added
- [x] Modified code has meaningful comments
- [x] All TODOs are linked with the issues
- [ ] All comments are resolved

## Testing
```
# Create pod with volumes
[root@servicenode ~]# kubectl get volumes
NAME                                       SIZE            STORAGE CLASS   HEALTH   CSI_STATUS   LOCATION                               NODE
pvc-49e60610-342c-4510-ae72-5060cd4dfe45   8001000000000   HDD             GOOD     CREATING     5898a63e-bcd0-4137-90b8-9a5671c0f9f8   212642c9-daa3-4e57-a055-8de60f774ff9
pvc-99b9c037-41cd-471a-aa2d-e293cf15dcf1   8001000000000   HDD             GOOD     CREATING     acf247e5-ce11-4252-b4d3-3de60d603075   212642c9-daa3-4e57-a055-8de60f774ff9
pvc-ebd45cce-d8e3-4057-bd09-f433c829183a   8001000000000   HDD             GOOD     CREATING     c809b14b-3caa-4917-a0e5-8f96a63811ef   212642c9-daa3-4e57-a055-8de60f774ff9

# Check wbt_lat_usec
[root@servicenode ~]# kubectl get drive -o wide 5898a63e-bcd0-4137-90b8-9a5671c0f9f8
NAME                                   SIZE            TYPE   HEALTH   STATUS   USAGE    SYSTEM   PATH       SERIAL NUMBER   NODE                                   SLOT
5898a63e-bcd0-4137-90b8-9a5671c0f9f8   8001000000000   HDD    GOOD     ONLINE   IN_USE            /dev/sds   VDH1950D        212642c9-daa3-4e57-a055-8de60f774ff9   5
[root@servicenode ~]# kubectl get csibmnode | grep 212642c9-daa3-4e57-a055-8de60f774ff9
csibmnode-212642c9-daa3-4e57-a055-8de60f774ff9   212642c9-daa3-4e57-a055-8de60f774ff9   worker0.ocp4.oil-bd.com   10.245.137.190

[root@servicenode ~]# ssh worker0
[core@worker0 ~]$ lsblk /dev/sds
NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
sds     65:32   0  7.3T  0 disk 
`-sds1  65:33   0  7.3T  0 part /var/lib/kubelet/pods/c5b8064b-b113-446d-bd76-43ae6da22bdf/volumes/kubernetes.io~csi/pvc-49e60610-342c-4510-ae72-5060cd4dfe45/mount
[core@worker0 ~]$ cat /sys/block/sds/queue/wbt_lat_usec
75000

# Edit Wbt Options
log: Dec 14 15:02:51.829 [INFO] [VolumeManager] Wbt config changed: {Enable:true Value:0 VolumeOptions:{Modes:[FS] StorageClasses:[]}}

[root@servicenode ~]# kubectl edit cm node-config
configmap/node-config edited

log: Dec 14 15:12:51.833 [INFO] [VolumeManager] Wbt config changed: {Enable:true Value:0 VolumeOptions:{Modes:[FS] StorageClasses:[csi-baremetal-sc-hdd csi-baremetal-sc-ssd csi-baremetal-sc-nvme]}}

# Restart pod 
[root@servicenode ~]# kubectl delete po web-0
pod "web-0" deleted

# Value changed
[core@worker0 ~]$ cat /sys/block/sds/queue/wbt_lat_usec
0

# Delete sts
[root@servicenode ~]# kubectl delete sts --all
statefulset.apps "web" deleted

# Value restored
[core@worker0 ~]$ cat /sys/block/sds/queue/wbt_lat_usec
75000

### CODE REVIEW

File: docs/proposals/custom-wbt-settings.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# Proposal: Support option to change WBT settings
+
+Last updated: 07.12.2021
+
+
+## Abstract
+
+CSI should have an opportunity to automate WBT (writeback throttling) disabling for specific Volumes.
```
Comment (mishoyama) [inline docs/proposals/custom-wbt-settings.md:8]: For specific SCs I guess and not disabling, but managing.

File: docs/proposals/custom-wbt-settings.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+## Background
+
+Some Storage Systems, which use Baremetal CSI, may have performance issues due to incorrect WBT settings.
+So it is specific for each device (sdb, sdc, ...), CSI should have an opportunity to control it on Volume or Pod creation stage.
+
+## Proposal
+
+#### Operator part
+
+Add parameters to csi-baremetal-deployment helm chart
```
Comment (mishoyama) [inline docs/proposals/custom-wbt-settings.md:19]: I think it should be in SC description

File: docs/proposals/custom-wbt-settings.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+- `WBTDisableOnHDDSC` -> set `WBTDisable=on` for `csi-baremetal-sc-hdd`
+- `WBTDisableOnSSDSC` -> set `WBTDisable=on` for `csi-baremetal-sc-ssd`
+- `WBTDisableOnNVMESC` -> set `WBTDisable=on` for `csi-baremetal-sc-nvme`
+
+#### Node Part
+
+- Node Stage ->
+ 1. Check SC Parameter for `WBTDisable=on`
```
Comment (mishoyama) [inline docs/proposals/custom-wbt-settings.md:28]: I suggest to have `wbt_lat_usec` and value set to 0 for disabling. User might want to play with the parameters.

File: docs/proposals/custom-wbt-settings.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+2. Scan current value (`cat /sys/block/<drive>/queue/wbt_lat_usec`)
+3. Restore default, if it's "0" (`echo "-1" > /sys/block/<drive>/queue/wbt_lat_usec`)
+
+## Rationale
+
+We could trigger WBT disabling via PVC annotation as for fake-attach feature. 
+
+## Compatibility
+
+WBT is supported only for 4.10 kernel version and above. Link - https://cateee.net/lkddb/web-lkddb/BLK_WBT.html
```
Comment (mishoyama) [inline docs/proposals/custom-wbt-settings.md:43]: Do we need to generate any events when WBT value cannot be set?

File: docs/proposals/custom-wbt-settings.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+Last updated: 07.12.2021
+
+
+## Abstract
+
+CSI should have an opportunity to automate WBT (writeback throttling) disabling for specific Volumes.
+
+## Background
+
+Some Storage Systems, which use Baremetal CSI, may have performance issues due to incorrect WBT settings.
```
Comment (elpaev) [inline docs/proposals/custom-wbt-settings.md:12]: should we mention that there is a [bug ](https://access.redhat.com/solutions/4904681) in WBT in some versions of kernel which has a recommended workaround in disabling this feature?  just a thought.

File: docs/proposals/custom-wbt-settings.md (hunk 1 above)
Comment (safronovD) [reply docs/proposals/custom-wbt-settings.md]: fixed

File: docs/proposals/custom-wbt-settings.md (hunk 2 above)
Comment (safronovD) [reply docs/proposals/custom-wbt-settings.md]: Refactored with using ConfigMap instead

File: docs/proposals/custom-wbt-settings.md (hunk 3 above)
Comment (safronovD) [reply docs/proposals/custom-wbt-settings.md]: fixed

File: docs/proposals/custom-wbt-settings.md (hunk 4 above)
Comment (safronovD) [reply docs/proposals/custom-wbt-settings.md:67]: fixed

File: docs/proposals/custom-wbt-settings.md (hunk 5 above)
Comment (safronovD) [reply docs/proposals/custom-wbt-settings.md]: added, thank you

File: docs/proposals/custom-wbt-settings.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+# Is some nodes are not in, user should add values manually via editting CM
+  acceptable_kernels.yaml: |-
+    node_kernel_versions:
+      - 4.18.0-193.65.2.el8_2.x86_64
+```
+
+#### Node Part
+
+- Node Stage ->
+ 1. Check WBT ConfigMap
```
Comment (mishoyama) [inline docs/proposals/custom-wbt-settings.md:49]: Do you suggest to read ConfigMap on each `NodeStageVolume` request?

**Comment by mishoyama** [comment]
@safronovD this fix must also be merged to release-1.0 branch.

File: docs/proposals/custom-wbt-settings.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  config.yaml: |-
+    enable: true
+    # Value to set (uint32), 0 means disabling
+    wbt_lat_usec_value: 0
+    acceptable_volume_options:
+    # Block volumes don't take any impact from WBT
+      modes:
+        - FS
+      # It is risky to change WBT settings for LVG Volumes
+      storage_types:
```
Comment (mishoyama) [inline docs/proposals/custom-wbt-settings.md:34]: Storage types or SC names?

Comment (mishoyama) [reply docs/proposals/custom-wbt-settings.md:34]: How will that work with the future feature `user configurable storage classes`?

File: docs/proposals/custom-wbt-settings.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+ 2. Scan current value (`cat /sys/block/<drive>/queue/wbt_lat_usec`)
+ 3. Set passed one, if it's not equal (`echo <value> > /sys/block/<drive>/queue/wbt_lat_usec`)
+ 4. Add `wbt-changed=yes` annotation on Volume
+ 5. Send `WBTValueSetFailed` Event if error  
+
+- Node UnStage ->
+1. Check `wbt-changed=yes` annotation on Volume
+2. Restore default (`echo "-1" > /sys/block/<drive>/queue/wbt_lat_usec`)
+3. Delete annotation
+4. Send `WBTValueRestoreFailed` Event if error
```
Comment (mishoyama) [inline docs/proposals/custom-wbt-settings.md:59]: Will you try to restore WBT when unmount command failed on Unstage?

File: docs/proposals/custom-wbt-settings.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+## Compatibility
+
+WBT is supported only for 4.10 kernel version and above. Link - https://cateee.net/lkddb/web-lkddb/BLK_WBT.html
+
+## Open issues (if applicable)
+
+ID | Name | Descriptions | Status | Comments
+---| -----| -------------| ------ | --------
+ISSUE-1 | Should we enable WBT disabling by default for specific kernel version? | We could describe the kernel version map, where it is necessary  |   |  The kernel version map is not clarified
+ISSUE-2 | Is it correct to switch WBT settings on Stage/UnStage? | Other candidates - CreateVolume/DeleteVolume, Publish/UnPublish |   |
```
Comment (mishoyama) [inline docs/proposals/custom-wbt-settings.md:74]: In Create/DeleteVolume case settings won't be persisted on node reboot. When volume is shared across the pods you will have to manipulate WBT multiple times on PublishVolume

File: pkg/node/volumemgr.go [hunk 10]
```diff
@@ -1226,3 +1234,96 @@ func (m *VolumeManager) overrideDriveHealth(drive *api.Drive, overriddenHealth,
...
+	drive, err := m.crHelper.GetDriveCRByVolume(vol)
+	if err != nil {
+		return "", err
+	}
+	if drive == nil {
+		return "", fmt.Errorf("drive %s is not found", vol.Spec.Location)
+	}
+
+	// expected device path - /dev/<device>
+	splitedPath := regexp.MustCompile(`[A-Za-z0-9]+`).FindAllString(drive.Spec.Path, -1)
```
Comment (mishoyama) [inline pkg/node/volumemgr.go:1312]: Why do we need this check?

File: pkg/node/wbt/operations/wbt_ops.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+	"strconv"
+	"strings"
+
+	"github.com/dell/csi-baremetal/pkg/base/command"
+)
+
+const (
+	wbtValuePath = "/sys/block/%s/queue/wbt_lat_usec"
+	// isWBTEnabledCmdImpl is a CMD to check current WBT value
+	checkWbtValueCmd = "cat " + wbtValuePath
```
Comment (mishoyama) [inline pkg/node/wbt/operations/wbt_ops.go:31]: Do we realy need to use system tools like `cat` and `echo` to read from or write to file? How about go library?

File: docs/proposals/custom-wbt-settings.md (hunk 6 above)
Comment (safronovD) [reply docs/proposals/custom-wbt-settings.md:49]: No, just check value, which was updated by watcher

File: docs/proposals/custom-wbt-settings.md (hunk 7 above)
Comment (safronovD) [reply docs/proposals/custom-wbt-settings.md:34]: I mean storage types like HDD, SSD here. But it called `StorageClass` in Volume Spec, it brings some confusing

Comment (safronovD) [reply docs/proposals/custom-wbt-settings.md:34]: For my opinion, if there are io problems with all disks on specified OS kernel version, why do we want to configure different values?

File: docs/proposals/custom-wbt-settings.md (hunk 8 above)
Comment (safronovD) [reply docs/proposals/custom-wbt-settings.md:59]: Yes, we will restore default on Unstage in any case

**Review by mishoyama — APPROVED** [review]

File: docs/proposals/custom-wbt-settings.md (hunk 7 above)
Comment (safronovD) [reply docs/proposals/custom-wbt-settings.md:34]: fixed

File: docs/proposals/custom-wbt-settings.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+      storage_types:
+        - HDD
+        - SSD
+        - NVME
+
+# The list of acceptable kernel versions
+# Is some nodes are not in, user should add values manually via editting CM
+  acceptable_kernels.yaml: |-
+    node_kernel_versions:
+      - 4.18.0-193.65.2.el8_2.x86_64
```
[the remark this answers is not in the document]
Comment (safronovD) [reply docs/proposals/custom-wbt-settings.md:43]: fixed

File: docs/proposals/custom-wbt-settings.md (hunk 8 above)
Comment (safronovD) [reply docs/proposals/custom-wbt-settings.md:59]: added in proposal

File: docs/proposals/custom-wbt-settings.md (hunk 9 above)
Comment (safronovD) [reply docs/proposals/custom-wbt-settings.md]: added in proposal

File: pkg/node/wbt/operations/wbt_ops.go (hunk 11 above)
Comment (safronovD) [reply pkg/node/wbt/operations/wbt_ops.go:32]: created TODO - https://github.com/dell/csi-baremetal/issues/653

File: pkg/node/volumemgr.go (hunk 10 above)
Comment (safronovD) [reply pkg/node/volumemgr.go:1329]: We must have only device name here without `/dev`

**Review by rkiselenko — APPROVED** [review]

**Pull request merged**